### PR TITLE
Check inconsistency of backward calculation and show its function name

### LINKED ISSUE #278: Show in which function type-inconsistency between data and grad occurs.
When we set `Variable.grad`, type consistency between `Variable.data` and `Variable.grad` is checked (as we did in #210 ). If this type check fails, it emits error message. But this message does not show in which function this inconsistency occurs. This is not preferable for debugging.

### PR DESCRIPTION
**Pull request by unnonouno**
fix #278

### CODE REVIEW

**Comment by delta2323** [comment]
I have two concerns and I do not have clear answers to them.
- We check the consistency between `data` and `grad` also in `variable` [here](https://github.com/pfnet/chainer/blob/master/chainer/variable.py#L113). They have some overlaps.
- But, these two consistency checks could not be equivalent when we accumulate gradient (e.g. [in this case](https://github.com/pfnet/chainer/blob/master/chainer/variable.py#L296)). In the original code, `gx` need not to be same `dtype` as `data`. Only `x.grad + gx` need to be same `dtype`. So by this PR, we impose additional constraint to the gradient.

**Comment by unnonouno** [comment]
We need to check `grad` also in `Variable` because users can set `grad` directly.

File: chainer/variable.py [hunk 1]
```diff
@@ -284,6 +284,20 @@ def add_cand(cand):
...
+                        'Type of data and grad mismatch in %s:\n%s != %s'
+                        % (func.label, type(x.data), type(gx)))
+                if gx.dtype != x.data.dtype:
+                    raise TypeError(
+                        'Dtype of data and grad mismatch in %s:\n%s != %s'
+                        % (func.label, x.data.dtype, gx.dtype))
+                if gx.shape != x.data.shape:
+                    raise ValueError(
+                        'Shape of data and grad mismatch in %s:\n%s != %s'
+                        % (func.label, x.data.shape, gx.shape))
```
Comment (beam2d) [inline chainer/variable.py]: Could you unify the type checking code here with that in the grad.setter? It seems redundant (except showing the function label). https://github.com/pfnet/chainer/blob/master/chainer/variable.py#L153

File: chainer/variable.py [hunk 2]
```diff
@@ -141,24 +162,8 @@ def grad(self):
...
     @grad.setter
     def grad(self, g):
-        error_msg = '''
-This error is occured in two cases. The first case is when the user manually
-sets the Variable.grad incorrectly. The second case is when some Function
-implementation has a bug. If you do not manually set the Variable.grad in your
-script, please report this error to the issue tracker with the stack trace,
-the information of your environment, and your script:
-https://github.com/pfnet/chainer/issues/new.
-'''
```
Comment (beam2d) [inline chainer/variable.py]: Is it OK to remove this message?

**Comment by beam2d** [comment]
LGTM!

**Pull request merged**